### PR TITLE
AH-96 add holocene-context related fields

### PR DESCRIPTION
**Pull request by djrochford**
This is in preparation for a NeOlith PR related to AH-96, which restores input-context functionality.

The fields on brandedUserProfile are my idea; however, the fields on the Dialog and Dialog state documents are things that were already referred to in the code, but were missing from the schema. It appears that a feature related to input-contexts was half-finished before we launched NeOliTh; I'm in the process of finishing it as part of AH-96.

A lot of changes are just my auto-formatter; the field I added to brandedUserProfiles is `holoceneContext`. The fields I added to dialog and dialogState are both called `contexts`.

Note that I'm using the `before.update` hook on the BrandedUserProfile document to time-stamp changes to my new field; I generalised the code in that hook a little bit while I was at it.

### CODE REVIEW

File: collections/dataWorkflows.js [hunk 1]
```diff
@@ -6,46 +6,59 @@ Dialogs.attachSchema(new SimpleSchema({
...
+    }),
+    optional: true
+  }
 }))
 
 DialogStates = new Mongo.Collection('dialogStates')
 
 DialogStates.attachSchema(new SimpleSchema({
   '_id': {type: String, optional: true},
+  'contexts': {type: [String], optional: true},
```
Comment (AnthonyChiavelli) [inline collections/dataWorkflows.js]: it seems like these optional fields sometime end up being tricky for us. is there a reason not to default it to en empty array?

File: collections/brandedUserProfile.js [hunk 2]
```diff
@@ -26,47 +28,59 @@ BrandedUserSchema = new SimpleSchema({
...
+  optional: true},
   created: fields.date(),
   collegeName: fields.string(o),
   college: fields.string(o),
   collegeId: fields.string({optional: false}),
   control: fields.bool(o),
   dob: fields.date(o),
   email: fields.string(o),
-  enrollmentId: fields.string(o), //pantherId for gsu
+  enrollmentId: fields.string(o), // pantherId for gsu
```
Comment (jeremysimms) [inline collections/brandedUserProfile.js:40]: Is this field GSU only? Wondering if it's different from CRM id or whatever we use for that typically.

File: collections/brandedUserProfile.js [hunk 3]
```diff
@@ -314,30 +349,38 @@ BrandedUserSchema = new SimpleSchema({
...
+const DATED_FIELDS = {
+  '_contactSettings.canText': '_contactSettings.canTextLastModified',
+  'holoceneContext.currentContext': 'holoceneContext.lastMessageDate'
+}
+
 BrandedUserProfiles.before.update((userId, doc, fieldNames, modifier) => {
-  if (modifier.$set && modifier.$set['_contactSettings.canText'] !== undefined) {
-    modifier.$set['_contactSettings.canTextLastModified'] = new Date()
-  }
-});
```
Comment (edanaher) [inline collections/brandedUserProfile.js:343]: Did you mean to remove this snippet?  I suspect an incorrectly resolved merge conflict.

**Review by edanaher — DISMISSED** [review]
In addition to the (I believe) accidental removed code I noted, shouldn't these fields be added to the README?

Plus, this is a perfect example of how multiple commits can be used to make review easier; if you had done one commit with formatting cleanups and one with your real changes, it would have been much easier to look at your changes and skim the formatting, rather than having to scroll through all the formatting changes checking to see if they were significant.

(This is one thing that really bugs me about our development in general; yes, it takes a bit longer to properly split out commits and/or rebase to get a clean history, but it results in faster and higher-quality code review, so I think that tradeoff is worth it.)

File: collections/brandedUserProfile.js [hunk 4]
```diff
@@ -314,30 +349,38 @@ BrandedUserSchema = new SimpleSchema({
...
 
 BrandedUserProfiles = new Mongo.Collection('brandedUserProfiles')
 BrandedUserProfiles.attachSchema(BrandedUserSchema)
 
+const DATED_FIELDS = {
+  '_contactSettings.canText': '_contactSettings.canTextLastModified',
+  'holoceneContext.currentContext': 'holoceneContext.lastMessageDate'
+}
+
 BrandedUserProfiles.before.update((userId, doc, fieldNames, modifier) => {
```
Comment (jeremysimms) [inline collections/brandedUserProfile.js:384]: Just out of curiosity, is there a way to look at the schema itself to determine which fields are date fields? If you did it this way you wouldn't have to maintain a separate collection of keys for fields of type `Date`

File: collections/brandedUserProfile.js (hunk 3 above)
Comment (jeremysimms) [reply collections/brandedUserProfile.js:343]: he just moved it to the DATED_FIELDS const

File: collections/brandedUserProfile.js (hunk 4 above)
Comment (jeremysimms) [reply collections/brandedUserProfile.js:384]: @djrochford  https://github.com/aldeed/simple-schema-js#defaultvalue

Comment (jeremysimms) [reply collections/brandedUserProfile.js:384]: ^You might be able to set a default in the schema and then run schema.clean()

Comment (jeremysimms) [reply collections/brandedUserProfile.js:384]: Possibly more relevant https://github.com/aldeed/simple-schema-js#autovalue

File: collections/brandedUserProfile.js (hunk 2 above)
Comment (djrochford) [reply collections/brandedUserProfile.js:40]: I am not super clear on the meaning of `enrollmentId` these days; I think we tend to map unique id columns that we aren't using as the offical unique id of the school to this field in imports. Is that right=, @AnthonyReid99 ? In any case, this PR isn't touching that.

File: collections/brandedUserProfile.js (hunk 3 above)
Comment (edanaher) [reply collections/brandedUserProfile.js:343]: Indeed, I'm blind.  Thanks for clarifying!

Comment (djrochford) [reply collections/brandedUserProfile.js:343]: Yeah, I did. There is code which should achieve the same thing in the same place; did you see that?

File: collections/brandedUserProfile.js (hunk 4 above)
Comment (djrochford) [reply collections/brandedUserProfile.js:384]: The collection is there to co-ordinate between fields that get changed and fields that record when that change occurs; not to, in general, keep track of date fields. I'm using it in the code below.

**Review by edanaher — COMMENTED** [review]
I'd still like to see these fields added to the README, but I'm not sure if I'll check in again today, so will withdraw my requested changes.

File: collections/brandedUserProfile.js (hunk 4 above)
Comment (djrochford) [reply collections/brandedUserProfile.js:384]: I've tried to make that clearer in the latest push.

File: collections/brandedUserProfile.js (hunk 2 above)
Comment (jeremysimms) [reply collections/brandedUserProfile.js:40]: Oh right i see that this is just a whitespace change. My apologies!

**Review by jeremysimms — APPROVED** [review]

File: collections/brandedUserProfile.js (hunk 2 above)
Comment (AnthonyChiavelli) [reply collections/brandedUserProfile.js:40]: @djrochford @jeremysimms it isn't really used for anything now. just a common field that schools supply us in data uploads

**Review by AnthonyChiavelli — APPROVED** [review]

File: collections/dataWorkflows.js (hunk 1 above)
Comment (djrochford) [reply collections/dataWorkflows.js]: Good call, I can do that. I agree that we should have as few optional fields as possible.

Comment (djrochford) [reply collections/dataWorkflows.js]: I'm afraid of adding non-optional fields because it might break validation somewhere.

**Pull request merged**